File: beginner/003/D.py
MOD = 1000000007
C = 30

def nCm_pascal():
    pascal = [[0 for _ in range(C*C)] for _ in range(C*C)]
    for i in range(1, C * C):
        pascal[0][0] = 1
        for j in range(0, C * C):
            if i == j and j == 0:
                pascal[i][j] = 1
            else:
                pascal[i][j] = (pascal[i-1][j] % MOD + pascal[i-1][j-1] % MOD) % MOD
    return pascal


def main():
    r, c = list(map(int, input().split(' ')))
    x, y = list(map(int, input().split(' ')))
    d, l = list(map(int, input().split(' ')))
    pascal = nCm_pascal()
    print(pascal[x*y][d] * (r - x + 1) * (c - y + 1) % MOD)

if __name__ == '__main__':
    main()

File: beginner/003/D.rs
const MOD: usize = 1000000007;
const C: usize = 30;

static mut PASCAL: [[usize; C*C]; C*C] = [[0; C*C]; C*C];

#[allow(unused_must_use)]
fn read_intpair() -> (usize, usize) {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input);
    let v: Vec<usize> = input
        .trim()
        .split_whitespace()
        .map(|x| x.parse().unwrap())
        .collect();
    (v[0], v[1])
}

#[allow(non_snake_case)]
fn nCm_pascal() {
    unsafe {
        PASCAL[0][0] = 1;
    }
    for i in 1..C*C {
        for j in 0..C*C {
            if i == j || j == 0 {
                unsafe {
                    PASCAL[i][j] = 1;
                }
            } else {
                unsafe {
                    PASCAL[i][j] = (PASCAL[i-1][j] % MOD + PASCAL[i-1][j-1] % MOD) % MOD;
                }
            }
        }
    }
}

fn main() {
    let (r, c) = read_intpair();
    let (x, y) = read_intpair();
    #[allow(unused_variables)]
    let (d, l) = read_intpair();
    nCm_pascal();
    unsafe {
        println!("{}", PASCAL[x*y][d] * (r - x + 1) * (c - y + 1) % MOD);
    }
}
